addons/bus/websocket.py

```diff
@@ -35,16 +35,24 @@
 
 
 MAX_TRY_ON_POOL_ERROR = 10
-DELAY_ON_POOL_ERROR = 0.03
+DELAY_ON_POOL_ERROR = 0.2
+# Keep 10% of the cursors to handle connection requests, rest can be used to
+# fetch notifications/process incoming messages.
+WS_CURSORS_COUNT = int(int(config.get("db_maxconn_gevent") or config["db_maxconn"]) * 0.9)
+WS_CURSORS_SEM = threading.Semaphore(WS_CURSORS_COUNT)
 
 
+@contextmanager
 def acquire_cursor(db):
     """ Try to acquire a cursor up to `MAX_TRY_ON_POOL_ERROR` """
-    for tryno in range(1, MAX_TRY_ON_POOL_ERROR + 1):
-        with suppress(PoolError):
-            return odoo.registry(db).cursor()
-        time.sleep(random.uniform(DELAY_ON_POOL_ERROR, DELAY_ON_POOL_ERROR * tryno))
-    raise PoolError('Failed to acquire cursor after %s retries' % MAX_TRY_ON_POOL_ERROR)
+    with WS_CURSORS_SEM:
+        for tryno in range(1, MAX_TRY_ON_POOL_ERROR + 1):
```

Comment:
I'm not sure I understand this part. If the `Semaphore` is not blocked, doesn't that imply that a connection is available? And the other way around: if the `Semaphore` is blocked, then we don't go in the 'retry' loop.

Reply:
I believe the PR you mentioned is indeed more appropriate. I'll close this one. As for the semaphore part, it's only used within this file so the semaphore could be acquired, but there is no guarantee that incoming HTTP requests won't take the cursors before the websocket does.